dask/array/reductions.py

```diff
@@ -1053,6 +1081,138 @@ def _nanargmax(x, axis, **kwargs):
+    dsk = {}
+    n_vals = len(prefix_vals)
+    level = 0
+    if n_vals >= 2:
+        # upsweep
+        stride = 1
+        stride2 = 2
+        while stride2 <= n_vals:
+            for i in range(stride2 - 1, n_vals, stride2):
+                new_vals = []
+                for index, left_val, right_val in zip(
+                    indices[i], prefix_vals[i - stride], prefix_vals[i]
+                ):
+                    key = base_key + index + (level, i)
+                    dsk[key] = (binop, left_val, right_val)
+                    new_vals.append(key)
+                prefix_vals[i] = new_vals
+            stride = stride2
+            stride2 *= 2
+            level += 1
+
+        # downsweep
+        # With `n_vals == 3`, we would have `stride = 1` and `stride = 0`, but we need
+        # to do a downsweep iteration, so make sure stride2 is at least 2.
+        stride2 = builtins.max(2, 2 ** ceil(log2(n_vals // 2)))
+        stride = stride2 // 2
+        while stride > 0:
+            for i in range(stride2 + stride - 1, n_vals, stride2):
+                new_vals = []
+                for index, left_val, right_val in zip(
+                    indices[i], prefix_vals[i - stride], prefix_vals[i]
+                ):
+                    key = base_key + index + (level, i)
+                    dsk[key] = (binop, left_val, right_val)
+                    new_vals.append(key)
+                prefix_vals[i] = new_vals
+            stride2 = stride
+            stride //= 2
+            level += 1
+
+    if full_indices:
+        for index in full_indices[0]:
+            dsk[base_key + index] = (
+                _compute_prefixscan0,
+                func,
+                (x.name,) + index,
+                axis,
+                dtype,
+            )
+        if method == "blelloch":
+            for indexes, vals in zip(drop(1, full_indices), prefix_vals):
+                for index, val in zip(indexes, vals):
+                    dsk[base_key + index] = (
+                        _compute_prefixscan,
+                        func,
+                        binop,
+                        val,
+                        (x.name,) + index,
+                        axis,
+                        dtype,
+                    )
+        else:  # method == "blelloch-split"
+            # e.g., compute `cumsum` before adding it to the sum of previous values
+            level_key = (level,)
+            for indexes, vals in zip(drop(1, full_indices), prefix_vals):
+                for index, val in zip(indexes, vals):
+                    final_key = base_key + index
+                    func_key = final_key + level_key
+                    dsk[func_key] = (
+                        _compute_prefixscan0,
+                        func,
+                        (x.name,) + index,
+                        axis,
+                        dtype,
+                    )
+                    dsk[final_key] = (binop, val, func_key)
```

Comment:
Note to self: this implements two variants of the "blelloch" method.  For example, the "blelloch-split" method computes the `cumsum` in a separate task before combining it to the sum of previous blocks.  The "blelloch" method performs these in one step.

What little benchmarking I've done suggests "blelloch" is better than "blelloch-split" for `cumsum` and `cumprod`.  This may not hold true under scrutiny, and "blelloch-split" may be useful for slower functions.

I'm deleting "blelloch-split" for now, and leaving this note for posterity.